### DISCUSSION TITLE
Unable to authorize renovate to fetch latest images on ghcr.io (github packages) registry on private package

### DISCUSSION THREAD

**Discussion by rslinckx** [renovatebot/renovate#8643] [Request Help] [answered]
I recently switched my packages to the new github registry ghcr.io instead of the older docker.pkg.github.com github packages registry. In both cases the packages are private and can only be accessed via a github token which i encrypted and setup in the config as password

The older registry worked fine, using a config like this: 
```
  "hostRules": [
    {
      "domainName": "docker.pkg.github.com",
      "username": "xxxx",
      "encrypted": {
        "password": "xxxxx"
      }
    }
]
```

I used the following to authenticate to the newer ghcr.io registry (only the domainName part has changed):
```
    {
      "domainName": "ghcr.io",
      "username": "xxx",
      "encrypted": {
        "password": "xxx"
      }
    }
```

Note that the actual config is using both like this, in case it matters:
```
  "hostRules": [
    {
      "domainName": "docker.pkg.github.com",
      "username": "xxx",
      "encrypted": {
        "password": "xxx"
      }
    },
    {
      "domainName": "ghcr.io",
      "username": "xxx",
      "encrypted": {
        "password": "xxxx"
      }
    }
]
```

Now when renovate runs i get the following error in the logs (for example in run #299102059) :
```
DEBUG: Datasource unauthorized
{
  "datasource": "docker",
  "lookupName": "ghcr.io/xxx/xxx/xxx/xxx",
  "url": "https://ghcr.io/v2/xxx/xxx/xxx/xxx/tags/list?n=10000"
}
DEBUG: Failed to look up dependency ghcr.io/xxx/xxx/xxx/xxx (ghcr.io/xxx/xxx/xxx/xxx)(packageFile="docker-compose.yml", dependency="ghcr.io/xxx/xxx/xxx/xxx")
```

I can't find any other issue or discussion about ghcr.io, so i'm not sure whether this is a problem with ghcr.io support in general, or just my configuration that's missing something, although i can't see what could be misconfigured?

**Comment by viceice** [comment]
Did you tried to login directly from local docker daemon? Are the credentials working?

**Reply by rslinckx** [reply]
Yes everything is working fine when using the docker CLI on both registries using the same credentials

**Reply by rslinckx** [reply]
The password is a github token with read:packages scope, although i tried also with wider scopes, but renovate won't auth correctly

**Reply by rslinckx** [reply]
Another interesting data-point, in the github interface where i have a list of my tokens, the token i use in the renovate config states that it was never accessed, as if renovate won't even try to authenticate at all

**Reply by viceice** [reply]
i know it works for public images on `ghcr.io`, did you checked the urls in log? are they right?

eg: should look like `https://ghcr.io/v2/org/image/tags/list?n=10000`

working sample url: `https://ghcr.io/v2/visualon/builder/tags/list` (need auth token from `https://ghcr.io/token` of cause)

**Reply by rslinckx** [reply]
in my case the path is org/something/something/something and the logs show the path as pasted above. This is in a docker-compose file, which i can run successfully, so the paths are fine, in theory.

**Comment by cubic3d** [comment]
Did you get it fixed? I also experience problems with ghcr.io and auth.

**Reply by rslinckx** [reply]
No still not working

**Comment by h-no** [comment]
I have the same problems.
I tried with:
```
  "hostRules": [
    {
      "matchHost": "ghcr.io",
      "username": "***",
      "encrypted": {
        "password": "***"
      }
    }
  ],
 ```
and
```
  "hostRules": [
    {
      "matchHost": "ghcr.io",
      "encrypted": {
        "username": "***",
        "password": "***"
      }
    }
  ],
 ```
 Results with:
 ```
DEBUG: Datasource unauthorized
{
  "datasource": "docker",
  "lookupName": "ghcr.io/x/y/z",
  "url": "https://ghcr.io/v2/x/y/z/tags/list?n=10000"
}
DEBUG: Failed to look up dependency ghcr.io/x/y/z (ghcr.io/x/y/z)(packageFile="xxx/kustomization.yaml", dependency="ghcr.io/x/y/z")
```

**Reply by h-no** [reply]
Verified that the username and PAT before encryption works using `docker login ghcr.io`

**Reply by h-no** [reply]
Troubleshooting a bit.. 

I read https://docs.docker.com/registry/spec/auth/token/#how-to-authenticate and following that guide I can curl the requests needed in my terminal like this:

1. Initial request to figure out auth (x/y/z is my image, i.e. organization/repo/imgname):
```
curl -i https://ghcr.io/v2/x/y/z/tags/list?n=10000
HTTP/2 401
content-type: application/json
www-authenticate: Bearer realm="https://ghcr.io/token",service="ghcr.io",scope="repository:x/y/z:pull"
date: Thu, 26 Aug 2021 13:13:35 GMT
content-length: 73
x-github-request-id: xxx

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required"}]}
```
2. Getting the token (password is a GitHub Personal Access Token):
```
curl -i -H "Authorization: Basic <base64encoded username:password>" "https://ghcr.io/token?service=ghcr.io&scope=repository:x/y/z:pull"
HTTP/2 200
content-type: application/json
docker-distribution-api-version: registry/2.0
date: Thu, 26 Aug 2021 13:23:15 GMT
content-length: 69
x-github-request-id: xxx

{"token":"BASE64TOKEN"}
```
3. Performing the original request with auth token from previous request:
```
curl -i -H "Authorization: Bearer BASE64TOKEN" "https://ghcr.io/v2/x/y/z/tags/list?n=10000"
HTTP/2 200
content-type: application/json
docker-distribution-api-version: registry/2.0
date: Thu, 26 Aug 2021 13:24:42 GMT
content-length: 846
x-github-request-id: xxx

{"name":"x/y/z","tags":["v0.0.1", ... ]}
```

So on the ghcr.io side, the API requests seem to work.

**Comment by h-no** [comment]
This is the same issue as https://github.com/renovatebot/renovate/issues/11446 👍

**Comment by dargmuesli** [comment, marked as answer]
I found the following to be working:

```json
{
  "matchHost": "ghcr.io",
  "hostType": "docker",
  "username": "your-user-name",
  "password": "***********"
}
```

or in the UI:

<img width="587" height="564" alt="image" src="https://github.com/user-attachments/assets/e4354fec-1a53-45b7-89f0-0fb00e91fe7b" />

**Comment by d4rkfella** [comment]
Having the same issue as other people in the discussion , i've set up the hostRule as suggested in the last comment, but still getting this error in the logs. 
```Failed to look up docker package ghcr.io/<username>/<image-name>```

Did anyone manage to get renovate to properly authenticate and look up images in private registries ?

**Reply by jamietanna** [reply]
@d4rkfella can you please raise a separate Discussion? We try to avoid bumping old threads
